From d5b47c94aa31cb0447207b298be6aac3d9bf16d9 Mon Sep 17 00:00:00 2001
From: kongweibin <kongweibin2@huawei.com>
Date: Fri, 13 Oct 2023 18:08:30 +0800
Subject: [PATCH 4/8] net, bpf: Add a writeable_tracepoint to
 inet_stream_connect

A trace point is added to the connection process. Theebpf program
can be mounted to modify the return value of the function. This is
mandatory for delaying the establishment of an ebpf link. After
the connection is complete, a message is returned immediately and
no unnecessary operation is performed.

Signed-off-by: kongweibin <kongweibin2@huawei.com>
---
 include/trace/events/sock.h | 14 ++++++++++++++
 net/ipv4/af_inet.c          |  1 +
 2 files changed, 15 insertions(+)

diff --git a/include/trace/events/sock.h b/include/trace/events/sock.h
index 905b151bc..841627e15 100644
--- a/include/trace/events/sock.h
+++ b/include/trace/events/sock.h
@@ -203,6 +203,20 @@ TRACE_EVENT(inet_sock_set_state,
 			show_tcp_state_name(__entry->newstate))
 );
 
+#undef NET_DECLARE_TRACE
+#ifdef DECLARE_TRACE_WRITABLE
+#define NET_DECLARE_TRACE(call, proto, args, size) \
+	DECLARE_TRACE_WRITABLE(call, PARAMS(proto), PARAMS(args), size)
+#else
+#define NET_DECLARE_TRACE(call, proto, args, size) \
+	DECLARE_TRACE(call, PARAMS(proto), PARAMS(args))
+#endif
+
+NET_DECLARE_TRACE(connect_ret,
+	TP_PROTO(int *err),
+	TP_ARGS(err),
+	sizeof(int));
+
 #endif /* _TRACE_SOCK_H */
 
 /* This part must be outside protection */
diff --git a/net/ipv4/af_inet.c b/net/ipv4/af_inet.c
index 4ddbfccfd..6c9b39e96 100644
--- a/net/ipv4/af_inet.c
+++ b/net/ipv4/af_inet.c
@@ -729,6 +729,7 @@ int inet_stream_connect(struct socket *sock, struct sockaddr *uaddr,
 	lock_sock(sock->sk);
 	err = __inet_stream_connect(sock, uaddr, addr_len, flags, 0);
 	release_sock(sock->sk);
+	trace_connect_ret(&err);
 	return err;
 }
 EXPORT_SYMBOL(inet_stream_connect);
-- 
2.33.0

